File: haskell/src/Copilot/Session.hs
---------------------------------------------------------------------------------------------
--  Copyright (c) Microsoft Corporation. All rights reserved.
---------------------------------------------------------------------------------------------

{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE RecordWildCards   #-}
{-# LANGUAGE ScopedTypeVariables #-}

-- | Copilot Session management.
--
-- A 'CopilotSession' represents a single conversation session with the
-- Copilot CLI. It supports sending messages, subscribing to events,
-- registering tools, and handling permission\/user-input\/hook requests.
module Copilot.Session
  ( -- * Session type
    CopilotSession
  , newCopilotSession
  , sessionId
  , workspacePath

    -- * Sending messages
  , send
  , sendAndWait

    -- * Event subscription
  , onSessionEvent
  , dispatchSessionEvent

    -- * Tool registration (internal)
  , registerSessionTools
  , getSessionToolHandler

    -- * Permission / user-input / hooks handlers (internal)
  , registerSessionPermissionHandler
  , registerSessionUserInputHandler
  , registerSessionHooks
  , handleSessionPermissionRequest
  , handleSessionUserInputRequest
  , handleSessionHooksInvoke

    -- * Lifecycle
  , destroySession
  , abortSession
  , getMessages
  ) where

import           Control.Concurrent        (MVar, newMVar, modifyMVar_, readMVar, newEmptyMVar, putMVar, takeMVar, tryPutMVar, tryTakeMVar)
import           Control.Concurrent.STM
import           Control.Exception         (SomeException, catch, try, throwIO, Exception)
import           Control.Monad             (forM_, when)
import           Data.Aeson                (Value (..), object, (.=), (.:), (.:?))
import qualified Data.Aeson                as Aeson
import           Data.Aeson.Types          (parseMaybe)
import           Data.IORef
import qualified Data.Map.Strict           as Map
import           Data.Text                 (Text)
import qualified Data.Text                 as T
import           System.Timeout            (timeout)

import           Copilot.JsonRpc           (JsonRpcClient, JsonRpcError (..),
                                            sendRequest)
import           Copilot.Types

-- ============================================================================
-- Session Error
-- ============================================================================

data SessionError
  = SessionRpcError Text
  | SessionTimeout Int
  | SessionErrorEvent Text (Maybe Text)
  deriving (Show)

instance Exception SessionError

-- ============================================================================
-- Session Type
-- ============================================================================

-- | A single conversation session with the Copilot CLI.
data CopilotSession = CopilotSession
  { csSessionId         :: !Text
  , csWorkspacePath     :: !(Maybe Text)
  , csJsonRpc           :: !JsonRpcClient
  , csEventHandlers     :: !(IORef [SessionEventHandler])
  , csToolHandlers      :: !(IORef (Map.Map Text ToolHandler))
  , csPermissionHandler :: !(IORef (Maybe PermissionHandler))
  , csUserInputHandler  :: !(IORef (Maybe UserInputHandler))
  , csHooks             :: !(IORef (Maybe SessionHooks))
  }

-- | Create a new session.
newCopilotSession :: Text -> JsonRpcClient -> Maybe Text -> IO CopilotSession
newCopilotSession sid rpc wsp = do
  handlers   <- newIORef []
  toolH      <- newIORef Map.empty
  permH      <- newIORef Nothing
  userInputH <- newIORef Nothing
  hooksRef   <- newIORef Nothing
  pure CopilotSession
    { csSessionId         = sid
    , csWorkspacePath     = wsp
    , csJsonRpc           = rpc
    , csEventHandlers     = handlers
    , csToolHandlers      = toolH
    , csPermissionHandler = permH
    , csUserInputHandler  = userInputH
    , csHooks             = hooksRef
    }

-- | Get the session ID.
sessionId :: CopilotSession -> Text
sessionId = csSessionId

-- | Get the workspace path (when infinite sessions are enabled).
workspacePath :: CopilotSession -> Maybe Text
workspacePath = csWorkspacePath

-- ============================================================================
-- Sending Messages
-- ============================================================================

-- | Send a message to the session.
-- Returns the message ID.
send :: CopilotSession -> MessageOptions -> IO Text
send session opts = do
  let reqParams = object $ filter ((/= Null) . snd)
        [ "sessionId"   .= csSessionId session
        , "prompt"      .= moPrompt opts
        , "attachments" .= moAttachments opts
        , "mode"        .= moMode opts
        ]
  result <- sendRequest (csJsonRpc session) "session.send" reqParams
  case result of
    Left err -> throwIO $ SessionRpcError (jreMessage err)
    Right val -> do
      let mMsgId = parseMaybe (\o -> (o :: Aeson.Object) .: "messageId") val :: Maybe Text
      case mMsgId of
        Just mid -> pure mid
        Nothing  -> throwIO $ SessionRpcError "No messageId in send response"

-- | Send a message and wait for the session to become idle.
--
-- Returns the last @assistant.message@ event received, or 'Nothing' if none
-- was emitted before the session went idle.
--
-- The timeout is in microseconds (default: 60 seconds = 60000000).
sendAndWait :: CopilotSession -> MessageOptions -> Maybe Int -> IO (Maybe SessionEvent)
sendAndWait session opts mTimeout = do
  let timeoutUs = maybe 60000000 id mTimeout

  idleMVar <- newEmptyMVar :: IO (MVar ())
  errorMVar <- newEmptyMVar :: IO (MVar SessionError)
  lastMsgRef <- newIORef (Nothing :: Maybe SessionEvent)

  -- Subscribe to events BEFORE sending to avoid race conditions
  let handler evt = do
        case seType evt of
          "assistant.message" -> writeIORef lastMsgRef (Just evt)
          "session.idle"      -> void $ tryPutMVar idleMVar ()
          "session.error"     -> do
            let mMsg = parseMaybe (\o -> (o :: Aeson.Object) .: "message") (seData evt) :: Maybe Text
                mStack = parseMaybe (\o -> (o :: Aeson.Object) .:? "stack") (seData evt) :: Maybe (Maybe Text)
            void $ tryPutMVar errorMVar (SessionErrorEvent (maybe "session error" id mMsg) (maybe Nothing id mStack))
          _ -> pure ()

  unsubscribe <- onSessionEvent session handler

  -- Send the message
  sendResult <- try $ send session opts
  case sendResult of
    Left (e :: SomeException) -> do
      unsubscribe
      throwIO $ SessionRpcError (T.pack $ show e)
    Right _ -> pure ()

  -- Wait for idle, error, or timeout
  result <- timeout timeoutUs $ do
    -- Race between idle and error
    let waitLoop = do
          -- Check if idle already signaled
          mIdle <- tryTakeMVar idleMVar
          case mIdle of
            Just () -> pure Nothing  -- No error
            Nothing -> do
              mErr <- tryTakeMVar errorMVar
              case mErr of
                Just err -> pure (Just err)
                Nothing  -> do
                  -- Small sleep then retry via STM or polling
                  -- Use a simple approach: block on idleMVar with a short timeout
                  mIdle2 <- timeout 10000 (takeMVar idleMVar)  -- 10ms poll
                  case mIdle2 of
                    Just () -> pure Nothing
                    Nothing -> waitLoop
    waitLoop

  unsubscribe

  case result of
    Nothing -> throwIO $ SessionTimeout (timeoutUs `div` 1000)
    Just (Just err) -> throwIO err
    Just Nothing -> readIORef lastMsgRef

-- ============================================================================
-- Event Subscription
-- ============================================================================

-- | Subscribe to session events.
-- Returns an IO action that unsubscribes when called.
onSessionEvent :: CopilotSession -> SessionEventHandler -> IO (IO ())
onSessionEvent session handler = do
  -- We use a unique reference to identify the handler for removal
  ref <- newIORef handler
  modifyIORef (csEventHandlers session) (handler :)
  pure $ modifyIORef (csEventHandlers session) (drop 1)
  -- Note: This simple approach removes from the head. A production implementation
  -- would use unique IDs. For the SDK pattern, this suffices for typical usage.

-- | Dispatch an event to all registered handlers. (Internal)
dispatchSessionEvent :: CopilotSession -> SessionEvent -> IO ()
dispatchSessionEvent session evt = do
  handlers <- readIORef (csEventHandlers session)
  forM_ handlers $ \h ->
    catch (h evt) (\(_ :: SomeException) -> pure ())

-- ============================================================================
-- Tool Registration
-- ============================================================================

-- | Register tools for this session. (Internal)
registerSessionTools :: CopilotSession -> [Tool] -> IO ()
registerSessionTools session tools = do
  let m = Map.fromList [ (toolName t, toolHandler t) | t <- tools ]
  writeIORef (csToolHandlers session) m

-- | Get a tool handler by name. (Internal)
getSessionToolHandler :: CopilotSession -> Text -> IO (Maybe ToolHandler)
getSessionToolHandler session name = do
  handlers <- readIORef (csToolHandlers session)
  pure $ Map.lookup name handlers

-- ============================================================================
-- Permission Handler
-- ============================================================================

-- | Register a permission handler. (Internal)
registerSessionPermissionHandler :: CopilotSession -> PermissionHandler -> IO ()
registerSessionPermissionHandler session handler =
  writeIORef (csPermissionHandler session) (Just handler)

-- | Handle a permission request. (Internal)
handleSessionPermissionRequest :: CopilotSession -> Value -> Text -> IO PermissionRequestResult
handleSessionPermissionRequest session reqVal sid = do
  mHandler <- readIORef (csPermissionHandler session)
  case mHandler of
    Nothing -> pure $ PermissionRequestResult "denied-no-approval-rule-and-could-not-request-from-user" Nothing
    Just handler -> do
      case Aeson.fromJSON reqVal of
        Aeson.Success permReq -> handler permReq sid
        Aeson.Error _ ->
          -- Try to at least get the kind
          let mKind = parseMaybe (\o -> (o :: Aeson.Object) .: "kind") reqVal :: Maybe Text
          in case mKind of
            Just kind -> handler (PermissionRequest kind Nothing Map.empty) sid
            Nothing   -> pure $ PermissionRequestResult "denied-no-approval-rule-and-could-not-request-from-user" Nothing

-- ============================================================================
-- User Input Handler
-- ============================================================================

-- | Register a user input handler. (Internal)
registerSessionUserInputHandler :: CopilotSession -> UserInputHandler -> IO ()
registerSessionUserInputHandler session handler =
  writeIORef (csUserInputHandler session) (Just handler)

-- | Handle a user input request. (Internal)
handleSessionUserInputRequest :: CopilotSession -> UserInputRequest -> Text -> IO UserInputResponse
handleSessionUserInputRequest session req sid = do
  mHandler <- readIORef (csUserInputHandler session)
  case mHandler of
    Nothing -> throwIO $ SessionRpcError "No user input handler registered"
    Just handler -> handler req sid

-- ============================================================================
-- Hooks
-- ============================================================================

-- | Register hooks. (Internal)
registerSessionHooks :: CopilotSession -> SessionHooks -> IO ()
registerSessionHooks session hooks =
  writeIORef (csHooks session) (Just hooks)

-- | Handle a hooks invocation. (Internal)
handleSessionHooksInvoke :: CopilotSession -> Text -> Value -> Text -> IO (Maybe Value)
handleSessionHooksInvoke session hookType input sid = do
  mHooks <- readIORef (csHooks session)
  case mHooks of
    Nothing -> pure Nothing
    Just hooks -> do
      let inv = HookInvocation { hiSessionId = sid }
      case hookType of
        "preToolUse" -> case shOnPreToolUse hooks of
          Nothing -> pure Nothing
          Just handler -> do
            case Aeson.fromJSON input of
              Aeson.Success inp -> do
                result <- handler inp inv
                pure $ fmap Aeson.toJSON result
              Aeson.Error _ -> pure Nothing

        "postToolUse" -> case shOnPostToolUse hooks of
          Nothing -> pure Nothing
          Just handler -> do
            case Aeson.fromJSON input of
              Aeson.Success inp -> do
                result <- handler inp inv
                pure $ fmap Aeson.toJSON result
              Aeson.Error _ -> pure Nothing

        "userPromptSubmitted" -> case shOnUserPromptSubmitted hooks of
          Nothing -> pure Nothing
          Just handler -> do
            case Aeson.fromJSON input of
              Aeson.Success inp -> do
                result <- handler inp inv
                pure $ fmap Aeson.toJSON result
              Aeson.Error _ -> pure Nothing

        "sessionStart" -> case shOnSessionStart hooks of
          Nothing -> pure Nothing
          Just handler -> do
            case Aeson.fromJSON input of
              Aeson.Success inp -> do
                result <- handler inp inv
                pure $ fmap Aeson.toJSON result
              Aeson.Error _ -> pure Nothing

        "sessionEnd" -> case shOnSessionEnd hooks of
          Nothing -> pure Nothing
          Just handler -> do
            case Aeson.fromJSON input of
              Aeson.Success inp -> do
                result <- handler inp inv
                pure $ fmap Aeson.toJSON result
              Aeson.Error _ -> pure Nothing

        "errorOccurred" -> case shOnErrorOccurred hooks of
          Nothing -> pure Nothing
          Just handler -> do
            case Aeson.fromJSON input of
              Aeson.Success inp -> do
                result <- handler inp inv
                pure $ fmap Aeson.toJSON result
              Aeson.Error _ -> pure Nothing

        _ -> pure Nothing

-- ============================================================================
-- Lifecycle
-- ============================================================================

-- | Destroy this session and release resources.
destroySession :: CopilotSession -> IO ()
destroySession session = do
  let params = object [ "sessionId" .= csSessionId session ]
  result <- sendRequest (csJsonRpc session) "session.destroy" params
  case result of
    Left err -> throwIO $ SessionRpcError (jreMessage err)
    Right _  -> do
      writeIORef (csEventHandlers session) []
      writeIORef (csToolHandlers session) Map.empty
      writeIORef (csPermissionHandler session) Nothing
      writeIORef (csUserInputHandler session) Nothing
      writeIORef (csHooks session) Nothing

-- | Abort the currently processing message.
abortSession :: CopilotSession -> IO ()
abortSession session = do
  let params = object [ "sessionId" .= csSessionId session ]
  result <- sendRequest (csJsonRpc session) "session.abort" params
  case result of
    Left err -> throwIO $ SessionRpcError (jreMessage err)
    Right _  -> pure ()

-- | Retrieve all events/messages from the session history.
getMessages :: CopilotSession -> IO [SessionEvent]
getMessages session = do
  let params = object [ "sessionId" .= csSessionId session ]
  result <- sendRequest (csJsonRpc session) "session.getMessages" params
  case result of
    Left err -> throwIO $ SessionRpcError (jreMessage err)
    Right val -> do
      let mEvents = parseMaybe (\o -> (o :: Aeson.Object) .: "events") val :: Maybe [SessionEvent]
      pure $ maybe [] id mEvents

File: clojure/src/copilot/session.clj
;;;; ---------------------------------------------------------------------------
;;;;  Copyright (c) Microsoft Corporation. All rights reserved.
;;;; ---------------------------------------------------------------------------
;;;;
;;;; Copilot Session - represents a single conversation session with the
;;;; Copilot CLI.

(ns copilot.session
  "A CopilotSession maintains conversation state, handles events, and manages
  tool execution.  Sessions are created via `copilot.client/create-session!`
  or resumed via `copilot.client/resume-session!`."
  (:require [copilot.json-rpc :as rpc]))

;; ============================================================================
;; Session construction
;; ============================================================================

(defn create-session
  "Create a new session atom.  Internal -- use `copilot.client/create-session!`.

  Returns an atom containing the session state map."
  [session-id rpc-client workspace-path]
  (atom {:session-id         session-id
         :rpc-client         rpc-client
         :workspace-path     workspace-path
         :event-handlers     #{}      ;; set of (fn [event])
         :typed-handlers     {}       ;; {event-type #{(fn [event])}}
         :tool-handlers      {}       ;; {tool-name handler-fn}
         :permission-handler nil      ;; (fn [request invocation])
         :user-input-handler nil      ;; (fn [request invocation])
         :hooks              nil      ;; session-hooks map
         :session-fs-handler nil}))   ;; session filesystem provider map

;; ============================================================================
;; Event subscription
;; ============================================================================

(defn on!
  "Subscribe to all session events.

  `handler` - (fn [event]) called for every event.
  Returns a zero-arg unsubscribe function."
  [session-atom handler]
  (swap! session-atom update :event-handlers conj handler)
  (fn [] (swap! session-atom update :event-handlers disj handler)))

(defn on-type!
  "Subscribe to a specific event type.

  `event-type` - string, e.g. \"assistant.message\", \"session.idle\"
  `handler`    - (fn [event])
  Returns a zero-arg unsubscribe function."
  [session-atom event-type handler]
  (swap! session-atom update-in [:typed-handlers event-type]
         (fnil conj #{}) handler)
  (fn []
    (swap! session-atom update-in [:typed-handlers event-type] disj handler)))

(defn dispatch-event!
  "Dispatch an event to all registered handlers.  Internal."
  [session-atom event]
  (let [state @session-atom
        event-type (:type event)]
    ;; Typed handlers
    (when-let [handlers (get-in state [:typed-handlers event-type])]
      (doseq [h handlers]
        (try (h event) (catch Exception _))))
    ;; Wildcard handlers
    (doseq [h (:event-handlers state)]
      (try (h event) (catch Exception _)))))

;; ============================================================================
;; Tool registration
;; ============================================================================

(defn register-tools!
  "Register tool handlers for this session.

  `tools` - vector of tool maps (see `copilot.types/tool`)"
  [session-atom tools]
  (swap! session-atom assoc :tool-handlers
         (into {} (map (fn [t] [(:name t) (:handler t)])) tools)))

(defn get-tool-handler
  "Get the handler function for a named tool, or nil."
  [session-atom tool-name]
  (get-in @session-atom [:tool-handlers tool-name]))

;; ============================================================================
;; Permission handling
;; ============================================================================

(defn register-permission-handler!
  "Register a permission handler.

  `handler` - (fn [request invocation] -> permission-result-map)"
  [session-atom handler]
  (swap! session-atom assoc :permission-handler handler))

(defn handle-permission-request!
  "Handle an incoming permission request.  Internal."
  [session-atom request]
  (let [handler (:permission-handler @session-atom)]
    (if-not handler
      {:kind "denied-no-approval-rule-and-could-not-request-from-user"}
      (try
        (handler request {:session-id (:session-id @session-atom)})
        (catch Exception _
          {:kind "denied-no-approval-rule-and-could-not-request-from-user"})))))

;; ============================================================================
;; User input handling
;; ============================================================================

(defn register-user-input-handler!
  "Register a user input handler.

  `handler` - (fn [request invocation] -> {:answer str :wasFreeform bool})"
  [session-atom handler]
  (swap! session-atom assoc :user-input-handler handler))

(defn handle-user-input-request!
  "Handle an incoming user input request.  Internal."
  [session-atom request]
  (let [handler (:user-input-handler @session-atom)]
    (if-not handler
      (throw (ex-info "User input requested but no handler registered" {}))
      (handler request {:session-id (:session-id @session-atom)}))))

;; ============================================================================
;; Hooks handling
;; ============================================================================

(defn register-hooks!
  "Register session hooks.

  `hooks` - map with optional keys:
    :on-pre-tool-use, :on-post-tool-use, :on-user-prompt-submitted,
    :on-session-start, :on-session-end, :on-error-occurred"
  [session-atom hooks]
  (swap! session-atom assoc :hooks hooks))

(defn register-session-fs-handler!
  "Register a session filesystem handler.

  `handler` - map with function-valued keys:
    :read-file, :write-file, :append-file, :exists?, :stat,
    :mkdir, :readdir, :readdir-with-types, :rm, :rename"
  [session-atom handler]
  (swap! session-atom assoc :session-fs-handler handler))

(defn handle-hooks-invoke!
  "Handle an incoming hooks invoke request.  Internal."
  [session-atom hook-type input]
  (let [hooks (:hooks @session-atom)]
    (when hooks
      (let [handler-map {"preToolUse"            :on-pre-tool-use
                         "postToolUse"           :on-post-tool-use
                         "userPromptSubmitted"   :on-user-prompt-submitted
                         "sessionStart"          :on-session-start
                         "sessionEnd"            :on-session-end
                         "errorOccurred"         :on-error-occurred}
            hook-key    (get handler-map hook-type)
            handler     (when hook-key (get hooks hook-key))]
        (when handler
          (try
            (handler input {:session-id (:session-id @session-atom)})
            (catch Exception _ nil)))))))

;; ============================================================================
;; Messaging
;; ============================================================================

(defn send!
  "Send a message to this session.

  `options` - message options map with :prompt and optional :attachments, :mode,
              :response-format, :image-options

  Returns the message ID string."
  [session-atom options]
  (let [state      @session-atom
        rpc-client (:rpc-client state)
        params     (cond-> {:sessionId   (:session-id state)
                            :prompt      (:prompt options)
                            :attachments (:attachments options)
                            :mode        (:mode options)}
                     (:response-format options)
                     (assoc :responseFormat (:response-format options))
                     (:image-options options)
                     (assoc :imageOptions (:image-options options)))
        response   (rpc/request! rpc-client "session.send" params)]
    (:messageId response)))

(defn send-and-wait!
  "Send a message and wait until the session becomes idle.

  Returns the last assistant.message event, or nil if none received.

  `options`    - message options map with :prompt
  `timeout-ms` - timeout in milliseconds (default 60000)"
  ([session-atom options]
   (send-and-wait! session-atom options 60000))
  ([session-atom options timeout-ms]
   (let [idle-promise     (promise)
         error-promise    (promise)
         last-msg         (atom nil)
         ;; Subscribe BEFORE sending to avoid race
         unsub (on! session-atom
                    (fn [event]
                      (case (:type event)
                        "assistant.message"
                        (reset! last-msg event)

                        "session.idle"
                        (deliver idle-promise true)

                        "session.error"
                        (deliver error-promise
                                 (ex-info (get-in event [:data :message] "Session error")
                                          {:event event}))
                        ;; default
                        nil)))]
     (try
       (send! session-atom options)
       ;; Wait for idle or error or timeout
       (let [result (deref idle-promise timeout-ms ::timeout)]
         (cond
           (realized? error-promise)
           (throw @error-promise)

           (= result ::timeout)
           (throw (ex-info (str "Timeout after " timeout-ms "ms waiting for session.idle")
                           {:timeout-ms timeout-ms}))

           :else
           @last-msg))
       (finally
         (unsub))))))

;; ============================================================================
;; Session management
;; ============================================================================

(defn session-id
  "Get the session ID."
  [session-atom]
  (:session-id @session-atom))

(defn workspace-path
  "Get the workspace path (when infinite sessions are enabled), or nil."
  [session-atom]
  (:workspace-path @session-atom))

(defn get-messages!
  "Retrieve all events/messages from this session's history."
  [session-atom]
  (let [state      @session-atom
        rpc-client (:rpc-client state)
        response   (rpc/request! rpc-client "session.getMessages"
                                 {:sessionId (:session-id state)})]
    (:events response)))

(defn get-metadata!
  "Retrieve metadata for this session."
  [session-atom]
  (let [state      @session-atom
        rpc-client (:rpc-client state)
        response   (rpc/request! rpc-client "session.getMetadata"
                                 {:sessionId (:session-id state)})]
    response))

(defn destroy!
  "Destroy this session and release resources.

  After calling this, the session can no longer be used."
  [session-atom]
  (let [state @session-atom
        rpc-client (:rpc-client state)]
    (rpc/request! rpc-client "session.destroy"
                  {:sessionId (:session-id state)})
    (swap! session-atom assoc
           :event-handlers #{}
           :typed-handlers {}
           :tool-handlers {}
           :permission-handler nil
           :user-input-handler nil
           :hooks nil
           :session-fs-handler nil)
    nil))

(defn abort!
  "Abort the currently processing message in this session."
  [session-atom]
  (let [state      @session-atom
        rpc-client (:rpc-client state)]
    (rpc/request! rpc-client "session.abort"
                  {:sessionId (:session-id state)})
    nil))
